tailcfg/tailcfg.go

```diff
@@ -1218,6 +1219,41 @@ type PingRequest struct {
 	IP netaddr.IP
 }
 
+// PingResponse provides result information for a TSMP or Disco PingRequest.
+// Typically populated from an ipnstate.PingResult used in `tailscale ping`.
+type PingResponse struct {
+	Type string // ping type, such as TSMP or disco.
+
+	IP       string // ping destination
+	NodeIP   string // Tailscale IP of node handling IP (different for subnet routers)
+	NodeName string // DNS name base or (possibly not unique) hostname
+
+	Err            string
+	LatencySeconds float64
+
+	// Endpoint is the ip:port if direct UDP was used.
+	// It is not currently set for TSMP pings.
+	Endpoint string
+
+	// DERPRegionID is non-zero DERP region ID if DERP was used.
+	// It is not currently set for TSMP pings.
+	DERPRegionID int
```

Comment:
so if we want to see if a disco ping resulted in a direct connection we'll need to send several PingRequests from control spaced out by a couple seconds to see if it eventually finds a path?

Reply:
yes, I suspect we'll need to make a number of adjustments in the protocol once we aim to cover all of the behaviors.

I've been debating myself between sending more full round-trips or adding a field equivalent to `--until-direct`, and whether I might at some point want to send more than one ping response per pong token. I had a code path in an early prototype in control, though the complexity was much higher so I cut back to focus on getting some results sooner.

In the interest of working through the end to end incrementally, I think I want to defer this decision for some follow-up, though that might be soon.

Reply:
SGTM. This is fine for now. Not a bad thing to send multiple pings anyway. Maybe that's how it'll even end up.